### PR TITLE
Translations chapter for the guide

### LINKED ISSUE #34: add translations chapter
Would be good to have documentation on how to contribute translations.

Could use material from [Translating R to your Language](https://contributor.r-project.org/translating-r-to-your-language/) tutorial.

Would help to support events like the Collaboration Campfires (https://github.com/r-devel/rcontribution/issues/11) as well as self-learning.

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #34

### CODE REVIEW

**Comment by SaranjeetKaur** [comment]
Hi @MichaelChirico 

This looks fantastic! - I have also been writing the same chapter as part of the GSoD 2022 project.
Would you mind if I add materials from my writeup here?

**Comment by tdhock** [comment]
Hi @MichaelChirico this looks like a good start. For non-experts on gettext files, it would be useful to see a figure (node and edge diagram, nodes for files, edges for people/programs that create those files) or example (list of all files relevant to a given domain, in the source tree) to explain the relationship of the various files (pot, po, mo).

**Comment by MichaelChirico** [comment]
> Would you mind if I add materials from my writeup here?

Please do!

> it would be useful to see a figure... or example... to explain the relationship of the various files 

Great point!

File: 13-translations.Rmd [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Translations
+
+This chapter covers internationalization in R, i.e., the display of messages in languages other than English.
```
Comment (SaranjeetKaur) [inline 13-translations.Rmd:3]:
```suggestion
It is possible to translate the messages in R, including the warning messages, error messages, and menu labels into different languages other than standard language. Depending on the version of R that you are using some of the languages might already be available while some might need work. This chapter covers internationalization in R, i.e., the display of messages in languages other than English.
```

File: 13-translations.Rmd [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+### .po files
+
+### .mo files
+
+.po files are plain text, but while helpful for human readers, this is inefficient for consumption by computers.
+The .mo format is a "compiled" version of the .po file optimized for retrieving messages when R is running.
+
+In R-devel, the conversion from .po to .mo is done by R-core -- you don't need to compile these files yourself.
+They are stored in the R sources at `./src/library/translations/inst` in various language-specific subdirectories.
+
```
Comment (SaranjeetKaur) [inline 13-translations.Rmd:60]:
```suggestion

### Some useful tools

For being able to translate the messages in R, the following tools could be used. The [Poedit](https://poedit.net/) provides a GUI editor for writing translations in several language. An [online keyboard](https://www.lexilogos.com/keyboard/index.htm) could also help facilitate the translations process.

```

File: 13-translations.Rmd [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+focus on coding while getting more natural errors/warnings in their native tongue.
+
+The [`gettext` manual](https://www.gnu.org/software/gettext/manual/index.html) is a more canonical reference for
+deep understanding of how `gettext` works. This chapter will just give a broad overview of how it works, with
+particular focus on how things work for R, with the goal of making it as low-friction as possible for developers
+and users to contribute new/updated translations.
+
+## How translations work
+
+Each of the default packages distributed with R (i.e., those found in `./src/library` such as `base`, `utils`,
```
Comment (SaranjeetKaur) [inline 13-translations.Rmd:18]:
```suggestion
There are two basic files with extensions `.pot` and `.po` that are usually required during the Translation process in R. The `.pot` files are the template files which contain the error messages, warnings, and other similar messages, in R. In the template `.pot` file these messages will be available in standard English against the placeholder `msgid "Standard English message is placed here"`. Below every `msgid` there will be a placeholder for the translated message called `msgstr ""` and it would always be empty (default) in the `.pot` file. The `msgstr` is to be filled in the corresponding `.po` file - to include the appropriate translation. Both the template `.pot` file and the translated `.po` file should stored in the same directory always. Each of the default packages distributed with R (i.e., those found in `./src/library` such as `base`, `utils`,
```

**Comment by SaranjeetKaur** [comment]
@MichaelChirico - I've added some stuff in the chapter

Also this could be placed as chapter 8, that is, after the Documenting R chapter and before the Testing pre release R versions chapter - what do you think?

**Comment by SaranjeetKaur** [comment]
Once the chapter number is decided, can we open this for review?

File: 13-translations.Rmd (hunk 3 above)
Comment (benubah) [reply 13-translations.Rmd:18]: Could you add a link to an example `.pot` and `.po` file from the SVN repo? May help curious readers quickly see what these files look like and what to expect.

File: 13-translations.Rmd (hunk 1 above)
Comment (MichaelChirico) [reply 13-translations.Rmd:3]: Took another pass, PTAL.

File: 13-translations.Rmd (hunk 2 above)
Comment (MichaelChirico) [reply 13-translations.Rmd:60]: I specifically avoided mentioning tools like Poedit because they are paid/proprietary software... does the guide have any official stance on implicit endorsement of non-open-source tooling?

File: 13-translations.Rmd (hunk 3 above)
Comment (MichaelChirico) [reply 13-translations.Rmd:18]: implementation question -- what sort of link would be appropriate? Is a relative link possible? Or should we be perma-linking a commit from the `r-devel/r-svn` repo?

File: 13-translations.Rmd (hunk 2 above)
Comment (SaranjeetKaur) [reply 13-translations.Rmd:60]: I don't think we want to endorse any non-open-source tooling. I did not realise it earlier! Let me exclude it from being mentioned in the guide.

Comment (MichaelChirico) [reply 13-translations.Rmd]: > there are also some tools available like `tools::checkPoFiles()`, `tools::update_pkg_po()`, ... Could these be useful for translators

My guess is "not much". Maybe we could use an advanced section where this comes up. But those functions are more useful for maintainers than for translators. Basically, those helpers are good for generating `.po` files, but `.po` files are the starting point for translators.

Comment (MichaelChirico) [reply 13-translations.Rmd]: > Does RStudio offer any specific features for working with .pot/.po?

Not that I know of.

Comment (MichaelChirico) [reply 13-translations.Rmd]: > Maybe write a sentence on naming conventions here or somewhere else?

I thought that was covered in the earlier paragraph:

> There are two exceptions to the basic pattern described above. The first is the domain for C messages produced
by the `base` package...

Let me know if we can improve the wording/flow.

Comment (MichaelChirico) [reply 13-translations.Rmd]: > Does the msginit look/behave the same way under Win?

Sorry I don't have access to any Windows machines to experiment...

**Comment by MichaelChirico** [comment]
Sorry, the history's a bit hard to follow now that some of the comments are marked as Outdated. If you think I'm missing addressing something, please ping again.

More progress here...

File: 08-translations.Rmd [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+Parsing, we get the following associations:
+
+ - the `0` entry corresponds to when a number equals 1 modulo 10 (i.e., 1, 11, 21, 31, ...) _except_ numbers equaling
+   11 modulo 100 (i.e., 11, 111, 211, 311, ...). Combining, that's 1, 21, 31, ..., 91, 101, 121, 131, ..., 191, ...
+ - the `1` entry corresponds to numbers at least 2 modulo 10 (2, 3, ..., 8, 9, 12, 13, 14, ...) and _either_
+   below 10 modulo 100 (0, 1, ..., 9, 100, 101, ..., 109, ...) _or_ exceeding 20 modulo 100 (21, 22, ..., 99). Combining,
+   that's 2, 3, ..., 9, 22, 23, ..., 29, 32, 33, ... 39, ..., 102, 103, ..., 109, 122, 123, ...
+ - The `2` entry corresponds to all other numbers, i.e. 0, 10, 11, 12, ..., 19, 20, 30, ..., 90, 100, 110, 111, 112, ...
+
+TODO(michaelchirico): How to discover a new `nplurals/plural`.
```
Comment (hturner) [inline 08-translations.Rmd:138]: Can we put these TODOS in HTML comments?

File: 08-translations.Rmd [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+TODO(michaelchirico): sprintf templates (`%s`) and redirections (`%1$s`)
+TODO(michaelchirico): fuzzy translations
+
+### .mo files
+
+.po files are plain text, but while helpful for human readers, this is inefficient for consumption by computers.
+The .mo format is a "compiled" version of the .po file optimized for retrieving messages when R is running.
+
+In R-devel, the conversion from .po to .mo is done by R-core -- you don't need to compile these files yourself.
```
Comment (hturner) [inline 08-translations.Rmd:147]: R-core -> R Core

File: 13-translations.Rmd (hunk 2 above)
Comment (MichaelChirico) [reply 13-translations.Rmd]: marking resolved since it's making a bit harder to follow the comment chain here, feel free to unresolve/add new comments on the latest version, thanks!

File: 13-translations.Rmd [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+as given and generate .po files from them to be sent to an R-core maintainer as a patch.
+
+To emphasize, this section is almost always not needed for contributing translations -- it is here for
+completeness and edification.
+
+### .po files
+
+### .mo files
+
+.po files are plain text, but while helpful for human readers, this is inefficient for consumption by computers.
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 13-translations.Rmd]: marking resolved since it's making a bit harder to follow the comment chain here, feel free to unresolve/add new comments on the latest version, thanks!

File: 08-translations.Rmd [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+Each of the default packages distributed with R (i.e., those found in `./src/library` such as `base`, `utils`,
+and `stats`) contains a `po` directory that is the central location for cataloguing/translating each package's
+messages.
+
+### .pot files
+
+The `.pot` file is a snapshot of the messages available in a given **domain**. A domain in R typically identifies
+a source package and a source language (either R or C/C++). For example, the file `R-base.pot`
+(found in the R sources in `./src/library/base/po`) is a catalogue of all messages produced by R code in the
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:28]: Should we prefer a link to the svn trunk, or the GH mirror trunk? Former is more official of course, latter comes with syntax highlighting/other tooling integration, web-navigable history/blame, etc...

File: 08-translations.Rmd [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+are menu labels in the GUI. Depending on the version of R that you are using, some of the languages might already
+be available while others may need work. R leverages the [`gettext`](https://www.gnu.org/software/gettext/) program
+to handle the conversion from English to arbitrary target languages.
+
+Having messages available in other languages can be an important bridge for R learners not confident in English --
+rather than learning two things at once (coding in R and processing diagnostic information in English), they can
+focus on coding while getting more natural errors/warnings in their native tongue.
+
+The [`gettext` manual](https://www.gnu.org/software/gettext/manual/index.html) is a more canonical reference for
+deep understanding of how `gettext` works. This chapter will just give a broad overview, with particular focus on
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:14]: done, thanks!

File: 08-translations.Rmd [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+a source package and a source language (either R or C/C++). For example, the file `R-base.pot`
+(found in the R sources in `./src/library/base/po`) is a catalogue of all messages produced by R code in the
+`base` package, while `stats.pot` (_viz._, `./src/library/stats/po`) is a catalogue of all messages produced
+by C code in the `stats` package.
+
+There are two exceptions to the basic pattern described above. The first is the domain for C messages produced
+by the `base` package. Well, technically, they are not "produced by the `base` package" in the normal sense
+that code in the `src` directory of base is not where these messages come from, but rather, the C code which
+is the fundamental backing of R itself (especially, but not exclusively, the C code under `./src/main`).
+Given this idiosyncrasy, the associated `.pot` file is `R.pot` and it is found in the `po` directory for `base`.
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:36]: Agree it's a bit tough to convey what I want to here without getting too far into the weeds about how R is organized as a codebase. Your suggestion is progress -- I've tweaked a bit further, PTAL.

File: 08-translations.Rmd [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+and `stats`) contains a `po` directory that is the central location for cataloguing/translating each package's
+messages.
+
+### .pot files
+
+The `.pot` file is a snapshot of the messages available in a given **domain**. A domain in R typically identifies
+a source package and a source language (either R or C/C++). For example, the file `R-base.pot`
+(found in the R sources in `./src/library/base/po`) is a catalogue of all messages produced by R code in the
+`base` package, while `stats.pot` (_viz._, `./src/library/stats/po`) is a catalogue of all messages produced
+by C code in the `stats` package.
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:30]: Great idea, much improved, thanks

File: 08-translations.Rmd [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+available for running R on Windows. These messages are stored in the `RGui.pot` domain, also in the `po`
+directory for `base`, and are most commonly derived from C code found in `./src/gnuwin32`. One reason to keep
+this domain separate is that it is only relevant to one platform (Windows). In particular, Windows has historically
+different character encodings, so that it made more sense for Windows developers to produce translations
+for Windows, since it is non-trivial for non-Windows users to test their translations for the Windows GUI.
+
+#### Generating .pot files
+
+For outside contributors, there's no need to update .pot files -- translators will typically take the R .pot files
+as given and generate .po files from them to be sent to an R-core maintainer as a patch.
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:48]: I had Michael Lawrence in mind here. But you're right that a more typical flow may be:

translators --> language-specific maintainer --> Michael L.

Is that worth clarifying here? Neither translators nor language maintainers need to worry about generating .pot...

File: 08-translations.Rmd [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+.po files are the most important artifacts for translators. They provide the (human-readable!) mapping between
+the messages as they appear in the source code and how the messages will appear to users in translated locales.
+
+#### Singular messages
+
+Most messages appear as `msgid`/`msgstr` pairs. The former gives the message at it appears in the code, while the
+latter shows how it should appear in translation. For example, here is an error in German (locale: `de`) informing
+the user that their input must be of class `POSIXt`
+([source](https://github.com/r-devel/r-svn/blob/c715d61cb74b3fee2d035faed9b258e86e420b75/src/library/base/po/R-de.po#L537-L538)):
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd:63]: Not 100% sure I follow how you see things here; I changed 'source' in the link to '.po file source', does that help?

File: 08-translations.Rmd (hunk 4 above)
Comment (MichaelChirico) [reply 08-translations.Rmd:138]: Yep, done. I left them as visible TODOs with the intention of finishing them before merging, but if we've passed MVP we can just hide them for now.

**Comment by MichaelChirico** [comment]
> The existing sections seem nearly done. It could be good to finishing tidying those and merge them in before filling in new sections as I think the description of the .po/.poedit files will already be useful for people that are starting to contribute via weblate (c.f. #68).

Great idea, better to have something ready to refer to now that the translations are picking up a lot more traffic. That can also help direct efforts on where there are remaining points of confusion.

File: 08-translations.Rmd (hunk 7 above)
[the remark this answers is not in the document]
Comment (hturner) [reply 08-translations.Rmd]: Yes, my first thought was that we should link to the official sources, but I like your later use of linking to highlighted lines and in general the GH mirror is easier to browse.

So, perhaps we can go with linking to the GH mirror whenever we just expect people to read the files in the browser (as throughout this chapter). To be very clear about it, we could add a footnote to the first link, to say that we are linking to the GH mirror in this chapter as it is a better interface for browsing the source files and link to the actual svn repo there (https://svn.r-project.org/R/trunk/)

File: 08-translations.Rmd [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+latter shows how it should appear in translation. For example, here is an error in German (locale: `de`) informing
+the user that their input must be of class `POSIXt`
+([.po file source](https://github.com/r-devel/r-svn/blob/c715d61cb74b3fee2d035faed9b258e86e420b75/src/library/base/po/R-de.po#L537-L538)):
+
+```
+msgid "'to' must be a \"POSIXt\" object"
+msgstr "'to' muss ein \"POSIXt\" Objekt sein"
+```
+
+The same message can also be found in
```
Comment (hturner) [inline 08-translations.Rmd:69]:
```suggestion
Most messages appear as `msgid`/`msgstr` pairs. The former gives the message as it appears in the code, while the
latter shows how it should appear in translation. For example, here is an error in German (locale: `de`) informing
the user that their input must be of class `POSIXt`

    msgid "'to' must be a \"POSIXt\" object"
    msgstr "'to' muss ein \"POSIXt\" Objekt sein"

See this in context in the [`R-de.po` source file](https://github.com/r-devel/r-svn/blob/c715d61cb74b3fee2d035faed9b258e86e420b75/src/library/base/po/R-de.po#L537-L538).

The same message can also be found in
```

File: 08-translations.Rmd (hunk 12 above)
[the remark this answers is not in the document]
Comment (hturner) [reply 08-translations.Rmd]: I've made a suggestion now to show what I meant - I removed the bit about the GH mirror following our discussion above.

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply 08-translations.Rmd]: got it, thanks!

**Comment by MichaelChirico** [comment]
@SaranjeetKaur / @hturner thanks for both your reviews; let me know when you're happy to merge this initial version.

File: 08-translations.Rmd [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+The [`gettext` manual](https://www.gnu.org/software/gettext/manual/index.html) is a more canonical reference for a
+deep understanding of how `gettext` works. This chapter will just give a broad overview, with particular focus on
+how things work for R, with the goal of making it as low-friction as possible for developers and users to contribute
+new/updated translations.
+
+## How translations work
+
+Each of the default packages distributed with R (i.e., those found in `./src/library` such as `base`, `utils`,
+and `stats`) contains a `po` directory that is the central location for cataloguing/translating each package's
```
Comment (bettinagruen) [inline 08-translations.Rmd:21]: I would suggest to also indicate that these packages have priority base, i.e.,

and `stats`) -> and `stats` and which have priority base)

File: 08-translations.Rmd [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+messages.
```
Comment (bettinagruen) [inline 08-translations.Rmd:22]: It might be good to add here what the `po` directory contains before continuing discussing different file formats.

File: 08-translations.Rmd [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+The `.pot` file is a snapshot of the messages available in a given **domain**. A domain in R typically identifies
+a source package and a source language (either R or C/C++). For example, the file `R-stats.pot`
+(found in the R sources in `./src/library/stats/po`) is a catalogue of all messages produced by R code in the
+`base` package, while `stats.pot` is a catalogue of all messages produced by C code in the `stats` package.
+
+There are two exceptions to the basic pattern described above. The first is the domain for messages produced by
+the C code which is the fundamental backing of R itself (especially, but not exclusively, the C code under
+./src/main). The associated .pot file is R.pot and it is found in `./src/library/base/po`. R-base.pot is a
+normal .pot file because base has a normal `R` directory.
```
Comment (bettinagruen) [inline 08-translations.Rmd:34]: Do you mean package `base`?

File: 08-translations.Rmd [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+Each of the default packages distributed with R (i.e., those found in `./src/library` such as `base`, `utils`,
+and `stats`) contains a `po` directory that is the central location for cataloguing/translating each package's
+messages.
+
+### .pot files
+
+The `.pot` file is a snapshot of the messages available in a given **domain**. A domain in R typically identifies
+a source package and a source language (either R or C/C++). For example, the file `R-stats.pot`
+(found in the R sources in `./src/library/stats/po`) is a catalogue of all messages produced by R code in the
```
Comment (bettinagruen) [inline 08-translations.Rmd:28]: This might be easier to follow by first stating that stats/po contains to pot files and then explaining which ones they are and what they contain / are here for.

File: 08-translations.Rmd [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+`base` package, while `stats.pot` is a catalogue of all messages produced by C code in the `stats` package.
+
+There are two exceptions to the basic pattern described above. The first is the domain for messages produced by
```
Comment (bettinagruen) [inline 08-translations.Rmd:31]: Do these exceptions all refer to files contained in the po directory of the base package?

File: 08-translations.Rmd [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+the C code which is the fundamental backing of R itself (especially, but not exclusively, the C code under
+./src/main). The associated .pot file is R.pot and it is found in `./src/library/base/po`. R-base.pot is a
+normal .pot file because base has a normal `R` directory.
+
+The second is the domain for the Windows R GUI, i.e., the text in the menus and elsewhere in the R GUI program
+available for running R on Windows. These messages are stored in the `RGui.pot` domain, also in the `po`
+directory for `base`, and are most commonly derived from C code found in `./src/gnuwin32`. One reason to keep
+this domain separate is that it is only relevant to one platform (Windows). In particular, Windows has historically
+different character encodings, so that it made more sense for Windows developers to produce translations
+for Windows, since it is non-trivial for non-Windows users to test their translations for the Windows GUI.
```
Comment (bettinagruen) [inline 08-translations.Rmd:41]: for Windows -> specifically for Windows

File: 08-translations.Rmd [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+#### Generating .pot files
+
+For outside contributors, there's no need to update .pot files -- translators will typically take the R .pot files
```
Comment (bettinagruen) [inline 08-translations.Rmd:45]: Do you want to consisently use mark-up for .pot and .po files? I.e., `.pot` and `.po`.

File: 08-translations.Rmd [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+Most messages appear as `msgid`/`msgstr` pairs. The former gives the message as it appears in the code, while the
+latter shows how it should appear in translation. For example, here is an error in German (locale: `de`) informing
+the user that their input must be of class `POSIXt`
+
+```
+msgid "'to' must be a \"POSIXt\" object"
+msgstr "'to' muss ein \"POSIXt\" Objekt sein"
+```
+
+See this in context in the [`R-de.po` source file](https://github.com/r-devel/r-svn/blob/c715d61cb74b3fee2d035faed9b258e86e420b75/src/library/base/po/R-de.po#L537-L538).
```
Comment (bettinagruen) [inline 08-translations.Rmd:68]: Why are you not linking to the SVN, i.e.,

https://svn.r-project.org/R/trunk/src/library/base/po/R-de.po

File: 08-translations.Rmd [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+```
+msgid "Warning message:\n"
+msgid_plural "Warning messages:\n"
+msgstr[0] "Įspėjantis pranešimas:\n"
+msgstr[1] "Įspėjantys pranešimai:\n"
+msgstr[2] "Įspėjančių pranešimų:\n"
+```
+
+This corresponds to the 3 different ways to pluralize words in Polish.
```
Comment (bettinagruen) [inline 08-translations.Rmd:112]: Lithuanian?

**Comment by SaranjeetKaur** [comment]
@all-contributors please add @MichaelChirico for content, doc, research

**Comment by SaranjeetKaur** [comment]
@all-contributors please add @bettinagruen for review, research, question

**Comment by SaranjeetKaur** [comment]
@all-contributors please add @nbenn for review, research, question

**Comment by SaranjeetKaur** [comment]
@all-contributors please add @benubah for content, review, research, question

**Comment by SaranjeetKaur** [comment]
@all-contributors please add @tdhock for review, research, question

**Comment by SaranjeetKaur** [comment]
All the open comments and reviews on this chapter which might need further dicussions and work are now opened as new issues (with links to the original comment or review). As of now, I'm merging this PR so that the translations chapter is up on the devguide. 

@MichaelChirico thank you for your contributions! :)
